gematria/granite/python/token_graph_builder_model.py

```diff
@@ -340,3 +368,72 @@ def _create_graph_network_modules(
             residual_connection=options.EnableFeature.BY_FLAG,
         ),
     )
+
+  def _make_batch_feed_dict(self) -> model_base.FeedDict:
+    feed_dict = super()._make_batch_feed_dict()
+
+    feed_dict[self._instruction_annotations] = (
+        self._batch_graph_builder.instruction_annotations
+    )
+    return feed_dict
+
+
+class TokenGraphBuilderModelNodeEmbed(snt.Embed):
+  """Extends `snt.Embed` to include instruction annotations in node embeddings.
+
+  Generates node embeddings normally, then replaces the last `num_annotation`
+  values of the embeddings corresponding to instructions with the annotation
+  values. The embeddings for other node types remain unchanged.
+  """
+
+  def __init__(
+      self,
+      instruction_annotations,
+      instruction_node_mask,
+      **kwargs,
+  ) -> None:
+    """Initializes node embeddings.
+
+    Args:
+      instruction_annotations: Tensor holding instruction level runtime
+        annotations as in `BasicBlockGraphBuilder`.
+      instruction_node_mask: As in `BasicBlockGraphBuilder`.
+      **kwargs: Additional arguments to be passed to the internal `snt.Embed`.
+    """
+    super().__init__(**kwargs)
+
+    # The number of annotations per instruction.
+    self.num_annotations = int(instruction_annotations.shape[1])
+
+    if self.num_annotations > self.embed_dim:
+      raise ValueError('num_annotations cannot be greater than embed_dim.')
+
+    self.instruction_annotations = instruction_annotations
+    self.instruction_node_mask = instruction_node_mask
+
+  def __call__(self, inputs):
+    embeddings = super().__call__(inputs)
+
+    if self.num_annotations == 0:
+      return embeddings
+
+    out = tf.concat(
+        [
+            tf.slice(
+                embeddings,
+                begin=[0, 0],
+                size=[-1, self.embed_dim - self.num_annotations],
+            ),
+            tf.tensor_scatter_nd_update(
+                tf.slice(
+                    embeddings,
+                    begin=[0, self.embed_dim - self.num_annotations],
+                    size=[-1, self.num_annotations],
+                ),
+                indices=tf.where(self.instruction_node_mask),
+                updates=self.instruction_annotations,
+            ),
```

Comment:
Did you test this with TF inference? The set of supported ops in TFLite is a bit limited, and more specialized ones (like scatters and gathers) might not be there.

If not, please add a TODO here and we should test it when possible (and maybe look for another way to implement it).

Reply:
I do remember testing TFLite inference after these changes, and I didn't see anything related to the graphdef having unsupported ops. I can see scatter and gather related ops in the TFLite MLIR dialect `tfl` as well - so it looks like we are safe here :)

Reply:
I don't think we have the export to TFLite on any testing workflow (and I've been surprised by what is _not_ supported in the past). But let's go with this and fix problems when we hit them.